### PR TITLE
Add attribute support for GCP pub/sub

### PR DESCRIPTION
**Pull request by trink**
- Move batching from the plugin into the library

### CODE REVIEW

File: gcp/common.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    case LSB_PB_STRING:
+      {
+        int tag = 0;
+        int wiretype = 0;
+        while (p && p < e) {
+          p = lsb_pb_read_key(p, &tag, &wiretype);
+          if (wiretype == LSB_PB_WT_LENGTH) {
+            long long len;
+            p = lsb_pb_read_varint(p, e, &len);
+            if (p && (len >= 0 || p + len <= e)) {
```
Comment (ameihm0912) [inline gcp/common.cpp]: is it possible here for `p + len` to overflow negative?

File: gcp/sandboxes/heka/input/gcp_pubsub.lua [hunk 2]
```diff
@@ -65,46 +68,91 @@ local err_msg = {
...
-                        err_msg.Payload = err
-                        err_msg.Fields.data = data
-                        pcall(inject_message, err_msg)
-                    end
+
+                local ok, err = pcall(decode, msgs[i][1], msg, true)
+                if not ok or err then
+                    err_msg.Payload = err
+                    err_msg.Fields.data = data
+                    pcall(inject_message, err_msg)
```
Comment (ameihm0912) [inline gcp/sandboxes/heka/input/gcp_pubsub.lua]: does any failure here need to be handled?

Comment (trink) [reply gcp/sandboxes/heka/input/gcp_pubsub.lua]: no, if we fail injecting the failure, we simply drop it

File: gcp/common.cpp (hunk 1 above)
Comment (trink) [reply gcp/common.cpp]: yes

**Pull request merged**